File: Lab/Basic-DiD/na.R
#**********************************************************************
# na.R
# Description: illustrating a no anticipation violation with diff-in-diff
#**********************************************************************

set.seed(20200403)
library(dplyr)
library(fixest)

# First create the states
states <- data.frame(state = 1:40)

# Generate 25 firms per state (total 1000 firms)
data <- states %>%
  slice(rep(1:n(), each = 25)) %>%
  group_by(state) %>%
  mutate(firms = runif(25, 0, 5)) %>%
  ungroup()

# Expand data for 4 years per firm
data <- data %>%
  slice(rep(1:n(), each = 4)) %>%
  group_by(state, firms) %>%
  mutate(year = 1990 + row_number() - 1,
         n = row_number()) %>%
  ungroup()

# Generate ID
data$id <- group_indices(data, state, firms)

# Define treatment group (upper half)
data <- data %>%
  mutate(group = ifelse(id >= 500, 1, 0))

# Correct post-treatment indicator (no anticipation satisfied)
data <- data %>%
  mutate(post = ifelse(year >= 1991, 1, 0))

# Incorrect post-treatment indicator (no anticipation violated)
data <- data %>%
  mutate(post_na = ifelse(year >= 1992, 1, 0))

# Data generating process
data <- data %>%
  mutate(e = rnorm(nrow(.), 0, 1),
         y0 = firms + n + e)

# Constant treatment effect
data <- data %>%
  mutate(y1_c = y0 + ifelse(year >= 1991, 10, 0))

# Dynamic treatment effect
data <- data %>%
  mutate(y1_d = case_when(
    year == 1991 ~ y0 + 10,
    year == 1992 ~ y0 + 20,
    year == 1993 ~ y0 + 30,
    TRUE ~ y0
  ))

# Treatment effects
data <- data %>%
  mutate(delta_c = y1_c - y0,
         delta_d = y1_d - y0)

summary(data$delta_c[data$year >= 1991 & data$group == 1])
summary(data$delta_d[data$year >= 1991 & data$group == 1])

# Treatment indicator
data <- data %>%
  mutate(d = ifelse(year >= 1991 & group == 1, 1, 0))

# Observed outcomes (switching equation)
data <- data %>%
  mutate(y_c = ifelse(d == 1, y1_c, y0),
         y_d = ifelse(d == 1, y1_d, y0))

# Aggregate causal parameters
att_c <- mean(data$delta_c[data$year >= 1991 & data$group == 1])
att_d <- mean(data$delta_d[data$year >= 1991 & data$group == 1])

cat("ATT (constant):", att_c, "\n")
cat("ATT (dynamic):", att_d, "\n")

# Correct specifications
cat("\nCorrect specifications:\n")
feols(y_c ~ group * post, data = data, vcov = "hc1") %>% summary()
feols(y_d ~ group * post, data = data, vcov = "hc1") %>% summary()

# Incorrect specifications (NA violated)
cat("\nIncorrect specifications (NA violated):\n")
feols(y_c ~ group * post_na, data = data %>% filter(year >= 1991), vcov = "hc1") %>% summary()
feols(y_d ~ group * post_na, data = data %>% filter(year >= 1991), vcov = "hc1") %>% summary()

File: Lab/Example-Code/na.py
#**********************************************************************
# na.py
# Description: illustrating a no anticipation violation with diff-in-diff
#**********************************************************************

import numpy as np
import pandas as pd
import statsmodels.formula.api as smf

np.random.seed(20200403)

# Create states and firms
states = np.repeat(np.arange(1, 41), 25)
firms = np.random.uniform(0, 5, len(states))

# Expand for years
data = pd.DataFrame({'state': states, 'firms': firms})
data = data.loc[data.index.repeat(4)].reset_index(drop=True)
data['year'] = np.tile(np.array([1990, 1991, 1992, 1993]), int(len(data)/4))
data['n'] = data.groupby(['state', 'firms']).cumcount() + 1

# Create id
data['id'] = data.groupby(['state', 'firms']).ngroup()

# Treatment group: upper half
data['group'] = (data['id'] >= 500).astype(int)

# Correct post-treatment indicator (no anticipation satisfied)
data['post'] = (data['year'] >= 1991).astype(int)

# Incorrect post-treatment indicator (no anticipation violated)
data['post_na'] = (data['year'] >= 1992).astype(int)

# Data generating process
data['e'] = np.random.normal(0, 1, len(data))
data['y0'] = data['firms'] + data['n'] + data['e']

# Constant treatment effect
data['y1_c'] = data['y0'] + np.where(data['year'] >= 1991, 10, 0)

# Dynamic treatment effect
data['y1_d'] = data['y0'] + np.select(
    [data['year'] == 1991, data['year'] == 1992, data['year'] == 1993],
    [10, 20, 30],
    default=0
)

# Treatment effects
data['delta_c'] = data['y1_c'] - data['y0']
data['delta_d'] = data['y1_d'] - data['y0']

print(data.loc[(data['year'] >= 1991) & (data['group'] == 1), ['delta_c', 'delta_d']].describe())

# Treatment indicator
data['d'] = ((data['year'] >= 1991) & (data['group'] == 1)).astype(int)

# Observed outcomes (switching equation)
data['y_c'] = np.where(data['d'] == 1, data['y1_c'], data['y0'])
data['y_d'] = np.where(data['d'] == 1, data['y1_d'], data['y0'])

# Aggregate causal parameters
att_c = data.loc[(data['year'] >= 1991) & (data['group'] == 1), 'delta_c'].mean()
att_d = data.loc[(data['year'] >= 1991) & (data['group'] == 1), 'delta_d'].mean()

print(f"ATT (constant): {att_c}")
print(f"ATT (dynamic): {att_d}")

# Correct specifications
print("\nCorrect specifications:")
model_c = smf.ols('y_c ~ group * post', data=data).fit(cov_type='HC1')
print(model_c.summary())

model_d = smf.ols('y_d ~ group * post', data=data).fit(cov_type='HC1')
print(model_d.summary())

# Incorrect specifications (NA violated)
print("\nIncorrect specifications (NA violated):")
data_na = data[data['year'] >= 1991]

model_c_na = smf.ols('y_c ~ group * post_na', data=data_na).fit(cov_type='HC1')
print(model_c_na.summary())

model_d_na = smf.ols('y_d ~ group * post_na', data=data_na).fit(cov_type='HC1')
print(model_d_na.summary())
